lua/starfall/libs_cl/midi.lua

```diff
@@ -0,0 +1,82 @@
+if not SF.Require("midi") then return function() end end
+
+--Midi Library
+-- @name midi
+-- @class library
+-- @libtbl midi_library
+SF.RegisterLibrary("midi")
+
+return function(instance)
+-- Only usable by the owner of the starfall chip
+if LocalPlayer() ~= instance.player then return end
+
+local midi_library = instance.Libraries.midi
+
+-- Event hook for midi
+-- @class hook
+-- @client
+-- @libtbl midi_library
+-- @param number time the exact systime which the event occured
+-- @param number command the command code of the event.  First 4 bits are the command code and last 4 are the channel
+-- @param varargs the different parameters of the midi event.  Can have 1 or more.
+SF.hookAdd("MIDI", "midi")
+
+-- Gets a table of all midi devices' ports.
+-- @return table table of all midi ports.
+function midi_library.getPorts()
+    return midi.GetPorts()
+end
+
+-- Opens the midi port to make it available to grab events from.  This must be called before the hook.
+-- @param number portNumber the midi port to open.
+-- @return string the name of the midi device opened at the given port.
+function midi_library.openPort(portNumber)
+    local portNumber = portNumber or 0
+    if midi.IsOpened(portNumber) then
+        midi.Close(portNumber)
+    end
+    return midi.Open(portNumber)
+end
+
+-- Checks if the midi port is currently opened.
+-- @return boolean if the port is open
+function midi_library.isPortOpened(portNumber)
+    return midi.IsOpened(portNumber)
+end
+
+-- Closes the midi port
```

Comment:
Since you check if the port is open before closing in closeAllPorts, is that needed here?

Reply:
changed around how the closeAllPorts function works.  The check is now within the closePort function itself.  It will error out if the port you are trying to close is not open, so a check is needed.